findtrieentries.go

```diff
@@ -0,0 +1,354 @@
+package veracity
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-merklelog/massifs"
+	"github.com/datatrails/go-datatrails-merklelog/mmr"
+	"github.com/google/uuid"
+	"github.com/urfave/cli/v2"
+)
+
+/**
+ * find trie Entries finds the trie entry associated with a given trie key
+ */
+
+const (
+	logTenantFlagName = "log-tenant"
+
+	logIDFlagName = "log-id"
+
+	domainFlagName = "domain"
+
+	appIDFlagName = "app-id"
+
+	asLeafIndexesFlagName = "as-leafindexes"
+
+	massifRangeStartFlagName = "massif-start"
+	massifRangeEndFlagName   = "massif-end"
+)
+
+var (
+	ErrIncompleteMassifRange = fmt.Errorf("error both %v and %v need to be set for a massif range, or both need to be omitted", massifRangeStartFlagName, massifRangeEndFlagName)
+	ErrNoLogTenant           = fmt.Errorf("error, cannot find log tenant, please provide either %v or %v", logIDFlagName, logTenantFlagName)
+)
+
+// logIDToLogTenant converts the log id to the log tenant
+func logIDToLogTenant(logID string) (string, error) {
+
+	// first get the byte representation of the hex
+	logIdBytes, err := hex.DecodeString(logID)
+	if err != nil {
+		return "", err
+	}
+
+	// we don't know if its a log version 0 log id or a log version 1 log id
+
+	// attempt log version 1 first
+
+	// log version 1 is the byte representation of the uuid part of the log tenant
+	logIdUUid, err := uuid.ParseBytes(logIdBytes)
+	if err != nil {
+
+		// assume its log version 0 if it can't be parsed as bytes
+		// which is just utf-8 bytes of the log tenant string
+		return string(logIdBytes), nil
+	}
+
+	// if we get here we know its log version 1, so make the log tenant from the uuid
+	logTenant := fmt.Sprintf("tenant/%s", logIdUUid.String())
+
+	return logTenant, nil
+}
+
+// findRangeForAllMassifs returns the start and end massif index for all massifs in the log for the given log tenant
+//
+// returns massif start index, massif end index and error.
+func findRangeForAllMassifs(massifReader MassifReader, logTenant string) (int64, int64, error) {
+
+	// get the head massif
+	// NOTE: this is expensive, if there is a better way to do this lets do it
+	headMassifContext, err := massifReader.GetHeadMassif(context.Background(), logTenant)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return 0, int64(headMassifContext.Start.MassifIndex), nil
+
+}
+
+// findTrieKeys searchs the log of the given log tenant for matches to the given triekeys
+// and returns the leaf indexes (trie indexes) of all the matches as well as the number of trie entries considered
+func findTrieKeys(log logger.Logger, massifReader MassifReader, logTenant string, massifStartIndex int64, massifEndIndex int64, massifHeight uint8, trieKeys ...[]byte) ([]uint64, uint64, error) {
+
+	// search in all the massifs
+	if massifStartIndex == -1 {
```

Comment:
start = -1 then start == 0

end = -1 then keep going until MassifNotFound ?

Reply:
Ahh i was looking for the err code MassifNotFound, i really didn't want to call the head massif :) 

thanks!

Reply:
done!

there is also a case of `BlobNotFound` from azblob we have to catch, but at least we don't have to get the headmassif!